src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -852,33 +852,58 @@ internal unsafe int GetHashCodeOfStringCore(ReadOnlySpan<char> source, CompareOp
+            // the solution is to try to fill the sort key in a temporary buffer of size equal 4 x string length
+
+            int guessedSize = (int)Math.Min((long)source.Length * 4, 1024 * 1024); // 1MB is the biggest array that can be rented from ArrayPool.Shared without memory allocation
+
+            if (TryGetSortKey(source, options, guessedSize, out int actualSortKeyLength, out int hash))
             {
-                int sortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, null, 0, options);
+                return hash; // fast path, the buffer was big enough
+            }
 
-                byte[]? borrowedArr = null;
-                Span<byte> span = sortKeyLength <= 512 ?
-                    stackalloc byte[512] :
-                    (borrowedArr = ArrayPool<byte>.Shared.Rent(sortKeyLength));
+            // the temporary buffer was too small but the actual sort key length was returned
+            if (actualSortKeyLength > 0 // 0 means internal error in ucol_getSortKey
+                && TryGetSortKey(source, options, actualSortKeyLength, out _, out hash)) 
+            {
+                return hash;
+            }
 
-                fixed (byte* pSortKey = &MemoryMarshal.GetReference(span))
-                {
-                    if (Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, sortKeyLength, options) != sortKeyLength)
-                    {
-                        throw new ArgumentException(SR.Arg_ExternalException);
-                    }
-                }
+            throw new ArgumentException(SR.Arg_ExternalException);
+        }
+        
+        private unsafe bool TryGetSortKey(ReadOnlySpan<char> source, CompareOptions options, int requestedSortKeyLength, out int actualSortKeyLength, out int hash)
+        {
+            Debug.Assert(requestedSortKeyLength > 0);
+            Debug.Assert(source.Length > 0);
 
-                int hash = Marvin.ComputeHash32(span.Slice(0, sortKeyLength), Marvin.DefaultSeed);
+            byte[]? borrowedArr = null;
+            Span<byte> sortKey = requestedSortKeyLength <= 512 ?
+                stackalloc byte[512] :
+                (borrowedArr = ArrayPool<byte>.Shared.Rent(requestedSortKeyLength));
 
-                // Return the borrowed array if necessary.
-                if (borrowedArr != null)
-                {
-                    ArrayPool<byte>.Shared.Return(borrowedArr);
-                }
+            fixed (char* pSource = &MemoryMarshal.GetReference(source))
+            fixed (byte* pSortKey = &MemoryMarshal.GetReference(sortKey))
+            {
+                actualSortKeyLength = Interop.Globalization.GetSortKey(_sortHandle, pSource, source.Length, pSortKey, sortKey.Length, options);
+            }
 
-                return hash;
+            if (actualSortKeyLength > 0 && actualSortKeyLength <= sortKey.Length)
```

Comment:
>actualSortKeyLength > 0 && actualSortKeyLength <= sortKey.Length [](start = 16, length = 64)

nit: is it worth it to cache the result of this condition 

bool ret = actualSortKeyLength > 0 && actualSortKeyLength <= sortKey.Length?

I don't know if the jit will optimize and not repeating evaluating the conditions at the end of the method.

Reply:
Does the function ever return anything other than the required length, or 0 in case of error?